### PR TITLE
Dynamic Movie/Show Images from TMDB

### LINKED ISSUE #25: Suggestion: Dynamic image
Hi, there is a project called "premid" and they found a way to use a dynamic cover in YTM and sound cloud,  I hope we can implement it. I think they  use URLs in place of imageKey  

[Their code](https://github.com/PreMiD/Presences/blob/main/websites/Y/YouTube%20Music/presence.ts)

Thanks

image

***Update***

Even Netflix that's would be cool

[Code](https://github.com/PreMiD/Presences/tree/main/websites/N/Netflix)

image

### PR DESCRIPTION
**Pull request by xBiei**
## 📑 Description
Well, because I hate the trakt logo and the other 2 logos for shows and movies because they look ugly and unrelated to whatever I'm watching, and discord now accepts urls in RPC images I figured we should use TMDB resources to improve the quality of this app.

Also, other than being able to use dynamic images, this PR won't change the dependencies at all.
And of course, if the app failed to use dynamic images, it'll fallback to the old ones.
Closes #25 

## ✅ Checks
- [x] **code:** My pull request adheres to the code style of this project
- [x] **docs:** Added / updated
- [x] **tests:** All the tests have passed

### CODE REVIEW

**Comment by DareFox** [comment]
Kinda awkward to have two same PR (#42) 😅

**Comment by rollingmoai** [comment]
Duplicate of #42

**Comment by xBiei** [comment]
Yeah I know lol
Both are doing the same thing so it probably doesn't matter which is used
I just PR-ed the one I use

**Comment by rollingmoai** [comment]
Have you even checked existing PRs before making yours? It's like reinventing the wheel 😅

**Comment by xBiei** [comment]
> Side note: Have you even checked existing PRs before making yours? It's like reinventing the wheel 😅

I did actually. Submitted it anyway. :3

**Comment by afonsojramos** [comment]
@xBiei @DareFox The way I see it, there are two clear distinctions between the two PRs.

The advantage to @DareFox's PR I suppose is a two way sword. Because it uses tmdb crate which makes the code cleaner, however we are adding a dependency for just two URLs.

This PR has caching for the image links, which I think is a great advantage, doesn't use any additional crates 

So, in the end, I think I will close #42, and move forward with this one.

What I would improve on would be to also have a "disposable" tmdb token for quick usage, what do you think?

I still need to further test this implementation however, especially to test the situations where the image is not found.

**Comment by xBiei** [comment]
> What I would improve on would be to also have a "disposable" tmdb token for quick usage, what do you think?

Yeah I agree, that would be good for the UX.

> I still need to further test this implementation however, especially to test the situations where the image is not found.

If there's no url response, it'll use the media's type as url which discord fetches from the uploaded images to the app in their portal. 

> Note: There are some clippy annotations that I would also fix.

I'll check that out later

**Comment by rollingmoai** [comment]
> What I would improve on would be to also have a "disposable" tmdb token for quick usage, what do you think?

Yes, that would be a good idea alongside with built-in client IDs for Trakt and Discord (#46)

**Comment by DareFox** [comment]
> The advantage to @DareFox's PR I suppose is a two way sword. Because it uses tmdb crate which makes the code cleaner, however we are adding a dependency for just two URLs.
> 
> This PR has caching for the image links, which I think is a great advantage, doesn't use any additional crates
> 
> So, in the end, I think I will close #42, and move forward with this one

I agree, dependency just for url fetch is kinda ambigous and this PR is better. Also caching could save potential problem with burner token limit usage

> What I would improve on would be to also have a "disposable" tmdb token for quick usage, what do you think?

It would be very cool, it will make less hassle with configuration

Also, I have one question with this PR (I can't test rn, because my setup is borked): Is poster cropped or has original size? 
If not cropped, then Discord crops it and adds blur to it. I fixed it with [this commit](https://github.com/afonsojramos/discrakt/pull/42/commits/82a157c17e5a2ee3c88d5107221ef34c50b23556), and maybe you could apply this fix here too

**Comment by xBiei** [comment]
> Also, I have one question with this PR (I can't test rn, because my setup is borked): Is poster cropped or has original size? 
> 
> If not cropped, then Discord crops it and adds blur to it. I fixed it with [this commit](https://github.com/afonsojramos/discrakt/pull/42/commits/82a157c17e5a2ee3c88d5107221ef34c50b23556), and maybe you could apply this fix here too

Yeah I noticed it is cropped and kinda blurry.
I'll check it out.

File: src/trakt.rs [hunk 1]
```diff
@@ -93,11 +97,67 @@ impl Trakt {
...
+                    Ok(response) => response,
+                    Err(_) => {
+                        log("Failed to get image from tmdb-api");
+                        return None;
+                    }
+                };
+
+                match response.into_json::<serde_json::Value>() {
+                    Ok(body) => {
+                        let image_url = format!("https://image.tmdb.org/t/p/w500{}", body["posters"][0].clone().get("file_path").unwrap().as_str().unwrap());
```
Comment (afonsojramos) [inline src/trakt.rs:116]:
```suggestion
                        let image_url = format!("https://image.tmdb.org/t/p/w600_and_h600_bestv2{}", body["posters"][0].clone().get("file_path").unwrap().as_str().unwrap());
```

File: src/trakt.rs [hunk 2]
```diff
@@ -93,11 +97,67 @@ impl Trakt {
...
```
Comment (afonsojramos) [inline src/trakt.rs:144]:
```suggestion
                        let image_url = format!("https://image.tmdb.org/t/p/w600_and_h600_bestv2{}", body["posters"][0].clone().get("file_path").unwrap().as_str().unwrap());
```

**Comment by xBiei** [comment]
Looks like I got too busy and forgot about this lol.
It should be fine now.

**Comment by afonsojramos** [comment]
Thanks!

**Pull request merged**